File: e2e/desktop/support/workspace-slot-ui.swift
import ApplicationServices
import Foundation

struct Input: Decodable { let processId: Int32; let windowId: String; let slotId: String; let action: String }
func fail(_ message: String) -> Never { FileHandle.standardError.write(Data(message.utf8)); exit(2) }
func attribute(_ element: AXUIElement, _ name: CFString) -> CFTypeRef? {
  var value: CFTypeRef?
  return AXUIElementCopyAttributeValue(element, name, &value) == .success ? value : nil
}
func text(_ element: AXUIElement, _ name: CFString) -> String { attribute(element, name) as? String ?? "" }
func children(_ element: AXUIElement) -> [AXUIElement] { attribute(element, kAXChildrenAttribute as CFString) as? [AXUIElement] ?? [] }
func descendants(_ root: AXUIElement) -> [AXUIElement] {
  var values = children(root), index = 0
  while index < values.count && values.count <= 4096 {
    let current = values[index]; index += 1
    let role = text(current, kAXRoleAttribute as CFString)
    if role != "AXWebArea" && role != "AXHTMLContent" { values.append(contentsOf: children(current)) }
  }
  guard values.count <= 4096 else { fail("Native workspace tree exceeds bound") }
  return values
}
guard CommandLine.arguments.count == 2,
  let data = CommandLine.arguments[1].data(using: .utf8),
  let input = try? JSONDecoder().decode(Input.self, from: data) else { fail("Invalid slot input") }
let application = AXUIElementCreateApplication(input.processId)
let windows = attribute(application, kAXWindowsAttribute as CFString) as? [AXUIElement] ?? []
let matches = windows.filter { text($0, kAXIdentifierAttribute as CFString) == "com.rionstudio.runtime.appkit-window.v1:\(input.windowId)" }
guard matches.count == 1 else { fail("Exact AppKit workspace window missing") }
let slots = descendants(matches[0]).filter { text($0, kAXIdentifierAttribute as CFString) == "workspace-slot-status:\(input.slotId)" }
guard slots.count <= 1 else { fail("Duplicate native slot status") }
if slots.isEmpty { print("{\"phase\":\"ready\"}"); exit(0) }
let slot = slots[0]
let phase = text(slot, kAXValueAttribute as CFString)
if input.action == "retry" {
  let buttons = descendants(slot).filter { text($0, kAXRoleAttribute as CFString) == "AXButton" }
  guard buttons.count == 1, (attribute(buttons[0], kAXEnabledAttribute as CFString) as? Bool) == true,
    AXUIElementPerformAction(buttons[0], kAXPressAction as CFString) == .success else { fail("Visible slot retry unavailable") }
}
var width = 0.0, height = 0.0
if let value = attribute(slot, kAXSizeAttribute as CFString), CFGetTypeID(value) == AXValueGetTypeID() {
  var size = CGSize.zero
  if AXValueGetValue(value as! AXValue, .cgSize, &size) { width = size.width; height = size.height }
}
let output: [String: Any] = ["phase": phase, "width": width, "height": height]
let json = try JSONSerialization.data(withJSONObject: output)
print(String(data: json, encoding: .utf8)!)

File: e2e/desktop/support/workspace-slot-ui.ts
import { $ } from "@wdio/globals";
import { execFile } from "node:child_process";
import { resolve } from "node:path";
import { promisify } from "node:util";
import { electronDesktopE2eProbe } from "./electron-driver";
import { focusVisibleMacosAppKitRuntime } from "./native-application-actions";
import { withWindowsRuntimeHost } from "./native-runtime-tabs";

const executeFile = promisify(execFile);
export async function workspaceSlotUi(input: {
  platform: "macos" | "windows"; windowId: string; tabId: string;
  slotId: string; mainWindowHandle: string; action?: "read" | "retry";
}): Promise<{ phase: string; width?: number; height?: number }> {
  if (input.platform === "macos") {
    const processId = (await electronDesktopE2eProbe()).processId;
    if (input.action === "retry") await focusVisibleMacosAppKitRuntime({ processId, windowId: input.windowId });
    const output = await executeFile("/usr/bin/xcrun", ["swift",
      resolve(import.meta.dirname, "workspace-slot-ui.swift"),
      JSON.stringify({ ...input, processId, action: input.action ?? "read" })
    ], { encoding: "utf8", timeout: 15_000 });
    return JSON.parse(output.stdout);
  }
  return withWindowsRuntimeHost(input.mainWindowHandle, input.tabId, async () => {
    const slot = await $(`[data-workspace-slot-status='${input.slotId}']`);
    if (!await slot.isExisting()) return { phase: "ready" };
    const phase = await slot.getAttribute("data-phase");
    const size = await slot.getSize();
    if (input.action === "retry") {
      const retry = await slot.$("button");
      await retry.waitForClickable({ timeout: 10_000 });
      await retry.click();
    }
    return { phase: phase!, width: size.width, height: size.height };
  }, input.windowId);
}
